File: triangles.py
from random import choices
from strassen_v1 import strassen


def graphGen(dim, p):
    """ Generates graph with dim vertices, edge probability p. """
    matrix = [choices([0, 1], [1-p, p], k=dim)
              for i in range(dim)]
    return matrix


def triangles(dim, p):
    """ Computes number of triangles in a graph using Strassen's algorithm. """
    graph = graphGen(dim, p)
    cubedGraph = strassen(strassen(graph, graph), graph)
    diagSum = 0
    # Add diagonal entries
    for i in range(dim):
        diagSum += cubedGraph[i][i]
    numTriangles = diagSum / 6
    return numTriangles


if __name__ == "__main__":
    # Edit dims, probabilities, numTrials here
    dim = 1024
    probs = [0.01]
    numTrials = 5

    # Compute average number of triangles for given parameters
    for p in probs:
        avg = 0
        for i in range(numTrials):
            avg += triangles(dim, p)
        avg = avg / numTrials
        print("Average number of triangles, probability %f: %f" % (p, avg))

File: c++/triangles.cpp
#include <array>
#include <chrono>
#include <iostream>
#include <random>

#include "strassen.cpp"

using namespace std;
using namespace std::chrono;

void graphGen(vector<int>& matrix, int dim, float p) {
    // Generates graph with dim vertices, edge probability p.

    // Define random number generator
    unsigned seed = chrono::system_clock::now().time_since_epoch().count();
    std::default_random_engine generator(seed);
    // std::default_random_engine generator;
    std::bernoulli_distribution distribution(p);

    // Fill in random values for matrices
    for (int i = 0; i < dim; i++) {
        for (int j = i; j < dim; j++) {
            if (i == j) {
                matrix[i * dim + j] = 0;
            } else {
                int edgeVal = distribution(generator);
                matrix[i * dim + j] = edgeVal;
                matrix[j * dim + i] = edgeVal;
            }
        }
    }
    return;
}

double triangles(vector<int>& matrix, vector<int>& output, int dim, float p) {
    // Computes number of triangles in a graph using Strassen's algorithm.
    const double divFactor = 6.0;

    // Generate graph with given probabilities
    graphGen(matrix, dim, p);

    // Compute A^3
    strassen(matrix, matrix, output, dim);
    strassen(matrix, output, output, dim);

    // Add diagonal entries
    double diagSum = 0.0;
    for (int i = 0; i < dim; i++) {
        diagSum = diagSum + (double) output[i * dim + i];
    }
    double numTriangles = (double) diagSum / (double) divFactor;
    return numTriangles;
}

int main() {
    // Edit dims, probabilities, numTrials here
    int dim = 1024;
    std::array<float, 5> probs = {0.01, 0.02, 0.03, 0.04, 0.05};
    double numTrials = 5;

    // Initialize adjacency, output matrices
    vector<int> matrix;
    matrix.resize(dim * dim);
    vector<int> output;
    output.resize(dim * dim);

    // Compute average number of triangles for given parameters
    for (int i = 0; i < probs.size(); i++) {
        double avg = 0;
        for (int j = 0; j < numTrials; j++) {
            avg += triangles(matrix, output, dim, probs[i]);
        }
        avg = avg / numTrials;
        cout << "Average number of triangles, probability " << probs[i] << ": "
             << avg << endl;
        cout << "Expected number of triangles: "
             << 178433024 * (probs[i] * probs[i] * probs[i]) << endl;
    }

    return 0;
}
